pkg/reconciler/tektonrun/tektonrun.go

```diff
@@ -0,0 +1,188 @@
+	"fmt"
+
+	buildv1alpha1 "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
+	"github.com/shipwright-io/build/pkg/config"
+	"github.com/shipwright-io/build/pkg/ctxlog"
+	tektonv1alpha1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	knativeapis "knative.dev/pkg/apis"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type ReconcileTektonRun struct {
+	config       *config.Config
+	client       client.Client
+	scheme       *runtime.Scheme
+	ownerRefFunc setOwnerReferenceFunc
+}
+
+func NewReconciler(c *config.Config, mgr manager.Manager, ownerRefFunc setOwnerReferenceFunc) reconcile.Reconciler {
+	return &ReconcileTektonRun{
+		config:       c,
+		client:       mgr.GetClient(),
+		scheme:       mgr.GetScheme(),
+		ownerRefFunc: ownerRefFunc,
+	}
+}
+
+func (r *ReconcileTektonRun) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	ctxlog.Debug(ctx, "received reconcile request for tekton run")
+	tektonRun := &tektonv1alpha1.Run{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: request.Namespace,
+			Name:      request.Name,
+		},
+	}
+	err := r.client.Get(ctx, client.ObjectKeyFromObject(tektonRun), tektonRun)
+	if errors.IsNotFound(err) {
+		ctxlog.Debug(ctx, "could not find tekton run, no status to update")
+		return reconcile.Result{Requeue: false}, nil
+	}
+	if err != nil {
+		ctxlog.Error(ctx, err, "failed to get tekton run")
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	if tektonRun.IsDone() {
+		ctxlog.Debug(ctx, "tekton run is done")
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	ctxlog.Info(ctx, "reconciling status for tekton run")
+
+	if err := ValidateTektonRun(tektonRun); err != nil {
+		return r.handleRunFailure(ctx, tektonRun, "InvalidRun", err)
+	}
+
+	var status ExtraFields
+	err = tektonRun.Status.DecodeExtraFields(&status)
+	if err != nil {
+		ctxlog.Error(ctx, err, "failed to decode extra status fields")
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	buildrun := &buildv1alpha1.BuildRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: tektonRun.Namespace,
+		},
+	}
+
+	if status.IsEmpty() {
+		buildrun, err = r.createBuildRun(ctx, tektonRun)
+		if err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+		ctxlog.Info(ctx, "creating builrun for tekton run", "buildrun", buildrun.GetName())
+
+		// recording the BuildRun name created using ExtraFields
+		fields := ExtraFields{BuildRunName: buildrun.GetName()}
+		if err = tektonRun.Status.EncodeExtraFields(&fields); err != nil {
+			ctxlog.Error(ctx, err, "failed to encode extra status fields", "tekton.run", tektonRun.GetName())
+			return reconcile.Result{Requeue: true}, err
+		}
+		now := metav1.Now()
+		tektonRun.Status.StartTime = &now
```

Comment:
Is it mandatory to set this here already? Because a created BuildRun is not yet started. My initial assumption would have been that the BuildRun's startTime would become the Run's startTime ? You actually do that in `handleRunFailure`, imo a no-op because the if nil check there will never be true if you set it here unconditionally.